### PR TITLE
Improve API error handling

### PR DESCRIPTION
**Pull request by jleach**
PoC work on improved API error handling. Opening PR for discussion.

### CODE REVIEW

**Comment by jleach** [comment]
TL;DR: I think its better to always use concrete types like `Error` where possible rather than ephemeral things like `Strings`.

That's a good question. Typically in Swift errors are not represented as "Strings" but rather an `Error` so consistency is important; so why invent a new way to deal with errors?

Also, your code you may want to do something different based on an Error (not always display it). For example an API call can generate two types of errors. The first, something went wrong with the network call and the Alamofire file request failed. Second, the request went through but the server didn't like it (Authentication failed, bla, bla). In the first case you probably want to retry one or two times before complaining to the user, the second you want to deal with it right away by triggering the Authentication workflow. By using and Error type consistently its makes it easy to do this in logic without string comparison:

comparisons. For example:
```
if (error.localisedDescription = "Something happened") {
  // do this
}
```

You want to do:
```
if (error == .sometingBadHappened) {
  // do this
}
```

What will probably happen eventually is the app will get converted to French because of government policy and we won't want to do string comparisons to deal with errors:
```if (error == "Authentification requise" || error == "Authentication required")```

**Comment by freshamir** [comment]
Makes sense now especially for when/if a different language is required. 

But the getAgreements endpoint still doesn't need to be changed that much. the completion handler can use APIError type instead of string. 

How can we implement this error handling without completely changing the api calls?

**Comment by jleach** [comment]
@freshamir I've added the code to pass on the error message to the user for. I didn't change much in the getAgreement; just enough to let it pass on the `APIError`. Check it out.
